### DISCUSSION TITLE
What is the "color" option default?

### DISCUSSION THREAD

**Discussion by stockiNail** [kurkle/chartjs-chart-sankey#52] [Ideas]
I have a doubt about what's the default of `color` option.

Having a look to the code and the `overrides` of the controller, I see:

https://github.com/kurkle/chartjs-chart-sankey/blob/95ed08b3f532e1f4a0180c549938363143f9999a/src/controller.js#L370

It's defined as scriptable options but it will be never called.

In fact, to set the `fillStyle` is using doing as following:

https://github.com/kurkle/chartjs-chart-sankey/blob/95ed08b3f532e1f4a0180c549938363143f9999a/src/controller.js#L208

where it seems that if `color` is missing, it uses `black`.

Furthermore, it's weird why is using `dataset.color` instead of `me.options.color`.
Doing some tests, using `me.options.color`, the callback to set the option seems to be invoked (but not the callback set in the `overrides`).

I don't know if it's an issue but I would like to use `me.options.color`. Make sense or am I wrong?

**Comment by kurkle** [comment]
I think you are correct, the `dataset` is not a "resolver", so it will not call the function. I think `me.options.color` is correct and should fallback to the dataset overrides, if the options is not defined. If it does not, then that is another issue to look at.

**Reply by stockiNail** [reply]
Very good, I'm gonna submit PR (but not now, sorry...)

**Reply by stockiNail** [reply]
@kurkle I did some tests.

I changed the following code from

https://github.com/kurkle/chartjs-chart-sankey/blob/b32c942d55c64a3cb401a45feebcb5d93cc981b5/src/controller.js#L216

to 

```javascript
ctx.fillStyle = me.options.color;
```

and using the [basic sample](https://github.com/kurkle/chartjs-chart-sankey/blob/master/samples/basic.html) to check the result.

The  `ctx.fillStyle` is set to `#666`, the default color option but is **NOT** triggering the the scriptable option set int the `overrides`:

https://github.com/kurkle/chartjs-chart-sankey/blob/b32c942d55c64a3cb401a45feebcb5d93cc981b5/src/controller.js#L378-L387

Then, having a look to the fallback of CHARTJS at [dataset level](https://www.chartjs.org/docs/latest/general/options.html#dataset-level-options), you can see:

### Dataset level options

`dataset.type` defaults to `config.type`, if not specified.

* dataset
* options.datasets[`dataset.type`]
* options
* overrides[`config.type`].datasets[`dataset.type`]
* defaults.datasets[`dataset.type`]
* defaults

and then the `overrides` are checked after the options, where the color option always exists because is the color for the chart for all other chart elements (axis, legend, title, and so on).

Therefore, in my opinion, new PR should:

1. remove line 384 of controller.js `color: () => '#efefef', `, because never used
2. apply `ctx.fillStyle = me.options.color;`

But I have a doubt: where should the dataset options default be set in the controller? `SankeyController.defaults`?

**Reply by stockiNail** [reply]
Another doubt. All options related to the nodes, leveraging on `me.options.*` are becoming to scriptable.
Nevertheless the context which is passed to the function is always a `dataset` context and doesn't allow to set different styles to each node but a option is used for all nodes. This could be the reason why the dataset instance is used to retrieve the options instead of the options of the controller, because at the end of the day the options can not be set for each node.
About color, you could want to have different label color for different nodes and that's not possible.

Having a look to the code, in order to activate different style for each node, the scriptable options for nodes need to have a specific context, where the node is part of the context otherwise is not possible to apply a consistent logic.

What do you think? Make sense?

**Reply by kurkle** [reply]
Been a while, but I think it makes sense. Somewhat related note, I think the nodes need to be made elements to let users interact with them.

**Comment by RishikeshB** [comment]
is there is any option to give background for label has shown in test sample from repo:
![column](https://github.com/kurkle/chartjs-chart-sankey/assets/94789179/e41863db-19a4-4102-9f35-ba0a8c8df95e)
